src/modem.rs

```diff
@@ -0,0 +1,373 @@
+            write!(f, "{:?}", self)
+        }
+    }
+
+    impl std::error::Error for ModemError {}
+
+    pub trait SimModem {
+        /// get the current mode of the sim modem
+        fn get_mode(&self) -> &CommunicationMode;
+
+        /// Initialise the remote modem so that it is in PPPoS mode.
+        fn negotiate(&mut self, comm: &mut UartDriver) -> Result<(), ModemError>;
+    }
+
+    impl From<usize> for ModemError {
+        fn from(value: usize) -> Self {
+            ModemError::ATBuild(value)
+        }
+    }
+
+    impl From<at_commands::parser::ParseError> for ModemError {
+        fn from(value: at_commands::parser::ParseError) -> Self {
+            ModemError::ATParse(value)
+        }
+    }
+
+    pub mod sim7600 {
+
+        use core::time::Duration;
+
+        use at_commands::{builder::CommandBuilder, parser::CommandParser};
+        use esp_idf_hal::{delay::TickType, uart::UartDriver};
+
+        use super::{CommunicationMode, ModemError, SimModem};
+        pub struct SIM7600(CommunicationMode);
+
+        impl SIM7600 {
+            pub fn new() -> Self {
+                Self(CommunicationMode::Command)
+            }
+        }
+
+        pub fn get_signal_quality<IO: embedded_svc::io::Read + embedded_svc::io::Write>(
+            comm: &mut IO,
+        ) -> Result<(i32, i32), ModemError> {
+            let mut buff = [0u8; 64];
+            let cmd = CommandBuilder::create_execute(&mut buff, true)
+                .named("+CSQ")
+                .finish()?;
+
+            comm.write(cmd).map_err(|_| ModemError::IO)?;
+
+            let len = comm.read(&mut buff).map_err(|_| ModemError::IO)?;
+
+            log::info!("got response {:?}", &buff[..len]);
+
+            // \r\n+CSQ: 19,99\r\n\r\nOK\r\n
+            Ok(CommandParser::parse(&buff[..len])
+                .expect_identifier(b"\r\n+CSQ: ")
+                .expect_int_parameter()
+                .expect_int_parameter()
+                .expect_identifier(b"\r\n\r\nOK\r\n")
+                .finish()?)
+        }
+
+        fn get_iccid<IO: embedded_svc::io::Read + embedded_svc::io::Write>(
+            comm: &mut IO,
+        ) -> Result<heapless::String<22>, ModemError> {
+            let mut buff = [0u8; 64];
+            let cmd = CommandBuilder::create_execute(&mut buff, true)
+                .named("+CICCID")
+                .finish()?;
+
+            comm.write(cmd).map_err(|_| ModemError::IO)?;
+
+            let len = comm.read(&mut buff).map_err(|_| ModemError::IO)?;
+            log::info!("got response {:?}", &buff[..len]);
+
+            let (ccid,) = CommandParser::parse(&buff[..len])
+                .expect_identifier(b"\r\n+ICCID: ")
+                .expect_raw_string()
+                .expect_identifier(b"\r\n\r\nOK\r\n")
+                .finish()?;
+
+            Ok(heapless::String::try_from(ccid).unwrap())
+        }
+
+        fn reset(comm: &mut UartDriver) -> Result<(), ModemError> {
```

Comment:
@ivmarkov 
Please note the difference between these two functions. I tried to make `SimModem` generic across `R,W`, but the read is giving me an issue.

One takes the `UartDriver` and one takes `embedded_io::Read+embedded_io::Write`

The blocking read in `embedded_io::Read` is never returning in my example code, and I have no idea why.

The timeout-based read in `UartDriver` works every time.

Reply:
I think there may be an issue with the impl of `embedded_io::Read` or a deeper fundamental problem in esp-idf with the uart, because this function sometimes does not return.
```
impl<'d> embedded_io::Read for UartDriver<'d> {
    fn read(&mut self, buf: &mut [u8]) -> Result<usize, Self::Error> {
        UartDriver::read(self, buf, delay::BLOCK).map_err(EspIOError)
    }
}
```

Reply:
It would only ever return when the other side sends at least one byte to you. Or else it will wait forever. You should only call it without a timeout (as you do) only when you expect an answer to something, or like a constant inflow of data.

Reply:
Whe you say the timeout based read works every time... are you actually checking the returned result? Because the only difference between a timeout read and a blocking read is that the blocking read will only return when it had read at least one Bute, while the timeout read might return when the timeout had expired. But under the hood, they are a single ESP IDF function?!

Reply:
https://github.com/esp-rs/esp-idf-hal/issues/475 See this ticket. I am sure there is an issue

Reply:
> When you say the timeout based read works every time... are you actually checking the returned result? Because the only difference between a timeout read and a blocking read is that the blocking read will only return when it had read at least one Bute, while the timeout read might return when the timeout had expired. But under the hood, they are a single ESP IDF function?!

Yes I do check the returned result. In this case, the buffer contains the value I expect from the modem, and I can parse it using `at_commands`. If the parse were failing (it hasn't happened yet), then I would be sure that the full message had not been read before the timeout.